### DISCUSSION TITLE
spi dependencies check must not have scope 'provided'

### DISCUSSION THREAD

**Discussion by Narcasserun** [apache/dolphinscheduler#6313] [Q&A]
-I'm doing the data source plug-in transformation now. I put the public code generated by the data source plug-in loading and connection object in the Dao module, but the default SPI dependencies check package check plug-in does not allow the provider scope dependency other than the SPI module. In my opinion, it's reasonable for the plug-in to rely on the scope of provide and will not package the Dao module.

-If the packaging check is a matter of principle, I can only separate a datasource API module based on Dao and let each data source plug-in introduce the datasource API module (connection pool, spring dependency + some general code), so that there will be datasource API dependency under each data source plug-in, just like the task plugin implementation, However, the scenario is obviously not very suitable for data source plug-in loading. The reason is that some data source related dependencies will be packaged and loaded repeatedly, because the API module and server module will load the environment (Dao module) that the data source plug-in depends on by default. And there will be a little too loose coupling

What do you think of this problem?

![image](https://user-images.githubusercontent.com/39329477/134474147-f5b91e33-0076-4752-bc0c-c6cb85e551e5.png)

Screenshot of the dependency of the sqlserver data source plug-in I implemented:
![image](https://user-images.githubusercontent.com/39329477/134473476-1262903f-4749-45bc-8697-40217b607040.png)
Screenshot of task shell plug-in dependency
![image](https://user-images.githubusercontent.com/39329477/134473904-16a6bc9d-42fc-418c-9c32-99cc28fbcbe6.png)

**Comment by ruanwenjun** [comment]
@Narcasserun For better discuss, please use English.

**Reply by Narcasserun** [reply]
I have expressed it in English
